### ISSUE TITLE
remux/clone to remove steganography/concatenated files

### ISSUE THREAD

**Issue by MayhemYDG** [jcupitt/libvips#244]
Is it possible for vips to remove data that's not part of the image data?
The goal is to disallow user-submitted images from containing concatenated files/steganography.
I believe no decoding/encoding is required for this, just copying of valid data.

I tried with
- `vips im_copy 1.jpg 2.jpg`: the result ended being a re-encoded file, and visually different.
- ImageMagick's `convert` tool: while the result was visually identical to the original, the file size increased.
- `ffmpeg -i 1.jpg -c:v copy -an -sn 2.jpg`: it didn't change anything, the concatenated data remained, and I've seen it re-encode some GIFs.

**Comment by jcupitt** [comment]
Hi, no, vips does not support copying without encode / decode. You can trim metadata with:

```
vips copy in.jpg out.jpg[strip]
```

and `out.jpg` will just have the minimum level of stuff attached.

I would look at the `cjpeg` program that comes with `libjpeg`. It supports copy without decompress / recompress and I believe it lets you edit metadata.

**Comment by MayhemYDG** [comment]
Thanks.
`cjpeg` gives me `Unrecognized input file format` for some reason.
I tried with `jpegtran` instead, and I got it working with `jpegtran -progressive in.jpg > out.jpg`.
Without `-progressive` the output would end up bigger, which leads me to believe I may be able to just use IM's `convert` tool if I figure out the proper settings, since I also need to do process PNG and GIF files.

**Comment by jcupitt** [comment]
ImageMagick will decompress and recompress, just like vips. 

You're right, it's `jpegtran`, try:

```
$ jpegtran -copy none k2.jpg > x.jpg
```

It should strip all metadata and not touch the image.

**Comment by MayhemYDG** [comment]
>ImageMagick will decompress and recompress, just like vips.

From my understanding vips will also decode and re-encode.
The result of `convert in.jpg out.jpg` is visually identical, unlike `vips copy in.jpg out.jpg`, unless I missed a configuration somewhere.

I've read about the jpegtran's `-copy` parameter, and I think I'll just use the default `-copy comments` or `-copy all`, since metadata isn't my main concern here, it's files that are the result of `cat file >> picture`.

Is such a feature out of scope for the vips project?

**Comment by jcupitt** [comment]
You can spot decode/recode by subtracting the two images and looking for the absolute maximum difference. If the compression has changed there will be a small difference.

For example:

```
john@katamata ~/pics $ convert k2.jpg -strip x.jpg
john@katamata ~/pics $ vips subtract k2.jpg x.jpg t1.v
john@katamata ~/pics $ vips abs t1.v t2.v
john@katamata ~/pics $ vips max t2.v
8.000000
```

ie. using `convert` you have a maximum absolute difference of 8. Whereas:

```
john@katamata ~/pics $ jpegtran -copy none k2.jpg > x.jpg
john@katamata ~/pics $ vips subtract k2.jpg x.jpg t1.v
john@katamata ~/pics $ vips abs t1.v t2.v
john@katamata ~/pics $ vips max t2.v
0.000000
```

Conversion is lossless with `jpegtran`.

**Comment by jcupitt** [comment]
Ah, I hadn't realised you were looking for data after the data, I thought you wanted to strip the image. That makes sense.

vips encodes with Q75, the default jpeg setting. You can set the encode Q factor with eg.:

```
vips copy k2.jpg x.jpg[Q=90]
```

**Comment by MayhemYDG** [comment]
Let me rephrase what I'm looking for:

Users can append anything at the end of a picture with something like `cat hidden >> picture` and then upload `picture`.
I want to remove the `hidden` part of `picture`, and keep `picture`.

I do not want the `picture` to be modified in any way:
- the file size shouldn't change if there was no `hidden` data.
- the file size must never increase.
- if it's progressive or not, it does not matter to me.
- I do not want to re-encode the image. The image cannot be altered.
- legitimate metadata can stay.
- should work for JPEG/PNG/GIF files.

I can sort of achieve that with `jpegtran -progressive -optimize`. It will force the picture to be progressive, which isn't too harmful. Problem is jpegtran only works for JPEGs.

**Comment by jcupitt** [comment]
Yes, I understand now. Sorry, vips will always decode/recode.

**Comment by MayhemYDG** [comment]
So if such a feature is out of scope for this project, do you know of any tool that solves my problem?
IM's `convert` parameters aren't always easy to comprehend for me.

**Comment by jcupitt** [comment]
It's a specialized job, for each file type you need a tool which has a very low-level understanding of the format. So jpegtran for jpg files, tiffsplit perhaps for tiff files, I don't know about png. 

Personally I think I would decode/recode, and perhaps remove metadata, since jpeg (and maybe others) make it easy to hide stuff in there as well.
